x-pack/legacy/plugins/apm/public/components/app/ServiceMap/Controls.tsx

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useContext, useState, useEffect } from 'react';
+import { EuiButtonIcon, EuiPanel } from '@elastic/eui';
+import theme from '@elastic/eui/dist/eui_theme_light.json';
+import styled from 'styled-components';
+import { i18n } from '@kbn/i18n';
+import { CytoscapeContext } from './Cytoscape';
+import { FullscreenPanel } from './FullscreenPanel';
+
+const Container = styled('div')`
+  left: ${theme.gutterTypes.gutterMedium};
+  position: absolute;
+  top: ${theme.gutterTypes.gutterSmall};
+`;
+
+const Button = styled(EuiButtonIcon)`
+  display: block;
+  margin: ${theme.paddingSizes.xs};
+`;
+
+const ZoomInButton = styled(Button)`
+  margin-bottom: ${theme.paddingSizes.s};
+`;
+
+const ZoomPanel = styled(EuiPanel)`
+  margin-bottom: ${theme.paddingSizes.s};
+`;
+
+const duration = parseInt(theme.euiAnimSpeedFast, 10);
+const steps = 5;
+
+function doZoom(cy: cytoscape.Core | undefined, increment: number) {
+  if (cy) {
+    const level = cy.zoom() + increment;
+    cy.animate({
+      duration,
+      zoom: { level, position: cy.$('.primary').position() }
+    });
+  }
+}
+
+export function Controls() {
+  const cy = useContext(CytoscapeContext);
+
+  const [zoom, setZoom] = useState((cy && cy.zoom()) || 1);
+
+  useEffect(() => {
+    if (cy) {
+      cy.on('zoom', event => {
```

Comment:
Is this event listener removed again? (to avoid memory leaks)

Reply:
Thanks for catching this. I forgot to put it there. It's been added in 799b5e9b9f9.

Reply:
I looked at that commit and I don't see anything related (it's reverting a mock change).
What I'd expect is the `useEffect` body to return a cleanup function that calls `cy.off('zoom', handler)` or something like it.

Reply:
Sorry no it was 799b5e9.